### PR TITLE
Changes for compatibility with ES 2.x

### PR DESCRIPTION
**Pull request by bignolip**
Changes to make plugin compatible with ES 2.x

### CODE REVIEW

**Comment by bignolip** [comment]
Still testing... Please wait until my next comment to merge.

File: pom.xml [hunk 1]
```diff
@@ -3,11 +3,11 @@
   <groupId>com.couchbase</groupId>
   <artifactId>elasticsearch-transport-couchbase</artifactId>
   <version>2.1.1</version>
-  
+
     <properties>
-        <elasticsearch.version>1.7.1</elasticsearch.version>
+        <elasticsearch.version>2.2.0-SNAPSHOT</elasticsearch.version>
```
Comment (st3v3nle) [inline pom.xml]: Should we use release version instead?

File: pom.xml [hunk 2]
```diff
@@ -25,7 +25,12 @@
             <version>${elasticsearch.version}</version>
             <scope>provided</scope>
         </dependency>
-
+        <dependency>
```
Comment (st3v3nle) [inline pom.xml]: Do we need this library?

File: pom.xml [hunk 3]
```diff
@@ -45,8 +56,8 @@
                 <artifactId>maven-compiler-plugin</artifactId>
                 <version>2.3.2</version>
                 <configuration>
-                    <source>1.6</source>
-                    <target>1.6</target>
+                    <source>1.8</source>
```
Comment (st3v3nle) [inline pom.xml]: Is this required to upgrade to 1.8?

Comment (bignolip) [inline pom.xml]: For the exact changes I made, yes, 1.8 is required.  Attempting to build the project with 1.7 or 1.6 will result in compile-time errors:

Requirement for 1.7:

In the org.elasticsearch.transport.couchbase.capi package, the PluginSecurity.java file imports java.nio.file.Path, which appears to be provided only in 1.7+.

Requirement for 1.8:

In the org.elasticsearch.transport.couchbase.capi package, the CouchbaseCAPITransportImpl class contains a method called "doStart".  Inside of "doStart", I added a "doPrivileged" block, in which a variable "portNumber"  is referenced.  The declaration of "portNumber" occurs in the surrounding declaration of "onPortNumber".  In 1.7-, the compiler cannot accept a reference to a non-final variable inside of the "doPrivileged" block.

However, if the declaration "public boolean onPortNumber(int portNumber)" is changed to "public boolean onPortNumber(final int portNumber)", there is no longer a need for 1.8, at least for compilation purposes.

If anything above doesn't make sense, please let me know and I will try to clarify.  Thanks!
--
If it is desirable to keep the project at 1.7 (or even 1.6), please let me know.  

I can easily make the changes to get compatibility with 1.7, but I'd like to test the project again.  

To get compatibility with 1.6, at the very least, the java.nio libraries will have to be included as a dependency.  I'm not sure if java.nio is only 1.7+.  There may be alternatives to java.nio as well (I think I just copied most of the "PluginSecurity" file from another ES open-source plugin project, so I haven't spent much time on it.).  

Please let me know how you would like me to proceed.  Thanks!

File: pom.xml (hunk 1 above)
Comment (bignolip) [inline pom.xml]: Is there a release version for ES 2.x or 2.2.0?

Actually, I see that I logged this as a fix for "Can't install Plugin couchbase 4.0 + elasticsearch 2.0", but that is no longer the case.

It turns out that there are considerable differences in getting compatibility with ES 2.x vs ES 2.0.  I originally thought I could roll them both into the same PR, but I cannot.  This PR is actually to ensure compatibility with ES 2.x, as acquired by following the instructions to build the 2.x branch of the ES GitHub repo.  

I am still working on putting together a PR for the compatibility fix for ES 2.0.

Please advise.

**Comment by bignolip** [comment]
One additional thing that I haven't looked into yet is the compatibility with other ES plugins, namely "marvel".  In my attempts to get the transporter to work with ES 2.0 (note that this PR is actually for ES 2.x), I noted that the ".%DATE%"-type indices created by "marvel" were configured to not allow dynamic mapping, which caused issues when the transporter tried to connect to them (during the couchbaseCheckpoint transport process). 

I'm not sure if there are 2.x-compatible versions of the other ES plugins, so I'm not sure if that is an issue for this particular PR.

I would add code to just filter out ES indices that begin with '.', but I'm not sure if that is going to have any repercussions with the XDCR contact, etc.  Any guidance on this issue would be appreciated.

File: src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java [hunk 4]
```diff
@@ -110,7 +133,8 @@ public CouchbaseCAPITransportImpl(Settings settings, RestController restControll
         this.bulkIndexRetryWaitMs = settings.getAsLong("couchbase.bulkIndexRetryWaitMs", 1000L);
         this.bucketUUIDCacheEvictMs = settings.getAsLong("couchbase.bucketUUIDCacheEvictMs", 300000L);
 
-        Class<? extends TypeSelector> typeSelectorClass = settings.<TypeSelector>getAsClass("couchbase.typeSelector", DefaultTypeSelector.class);
+        Class<? extends TypeSelector> typeSelectorClass = DefaultTypeSelector.class;
```
Comment (Branor) [inline src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java]: Why did you change this to always use the DefaultTypeSelector class, instead of whatever is specified in the config file?

File: src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java [hunk 5]
```diff
@@ -119,7 +143,7 @@ public CouchbaseCAPITransportImpl(Settings settings, RestController restControll
         this.typeSelector.configure(settings);
         logger.info("Couchbase transport is using type selector: {}", typeSelector.getClass().getCanonicalName());
 
-        Class<? extends ParentSelector> parentSelectorClass = settings.<ParentSelector>getAsClass("couchbase.parentSelector", DefaultParentSelector.class);
+        Class<? extends ParentSelector> parentSelectorClass = DefaultParentSelector.class;
```
Comment (Branor) [inline src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java]: Same question as for the type selector.

File: src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java [hunk 6]
```diff
@@ -128,7 +152,7 @@ public CouchbaseCAPITransportImpl(Settings settings, RestController restControll
         this.parentSelector.configure(settings);
         logger.info("Couchbase transport is using parent selector: {}", parentSelector.getClass().getCanonicalName());
 
-        Class<? extends KeyFilter> keyFilterClass = settings.<KeyFilter>getAsClass("couchbase.keyFilter", DefaultKeyFilter.class);
+        Class<? extends KeyFilter> keyFilterClass = DefaultKeyFilter.class;
```
Comment (Branor) [inline src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java]: And same as the type/parent selector here. Why only using the default?

File: src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java (hunk 5 above)
Comment (bignolip) [inline src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java]: @Branor In ES 2.0 and 2.x, there is no longer support for loading class names from Strings.  Settings objects no longer have a visible ClassLoader either.  The following code I think works around this problem, but I don't know if the "local" ClassLoader to the Impl class is guaranteed to be the right one:

This private method is added to the Impl class:

    @SuppressWarnings({"unchecked"})
    private <T> Class<? extends T> getAsClass(
      String className, Class<? extends T> defaultClazz) {
      String sValue = className;
      if (sValue == null) {
         return defaultClazz;
      }
      try {
         return (Class<? extends T>) this.getClass().getClassLoader().loadClass(sValue);
      } catch (ClassNotFoundException e) {
         throw new NoClassSettingsException("Failed to load class setting [" 
            + className + "] with value [" + sValue + "]", e);
      }
    }

The line in question is changed to:

    Class<? extends TypeSelector> typeSelectorClass = this.getAsClass(settings.get("couchbase.typeSelector"), DefaultTypeSelector.class);

I'll commit these changes to the PR if you think they are a suitable alternative to calling the now missing Settings.getAsClass method.

File: src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java (hunk 6 above)
Comment (bignolip) [inline src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java]: @Branor See the comment above.

If that change is valid, then this line would become:

    Class<? extends ParentSelector> parentSelectorClass = this.getAsClass(settings.get("couchbase.parentSelector"), DefaultParentSelector.class);

File: src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java (hunk 4 above)
Comment (bignolip) [inline src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java]: @Branor In ES 2.0 and 2.x, there is no longer support for loading class names from Strings.  Settings objects no longer have a visible ClassLoader either.  The following code I think works around this problem, but I don't know if the "local" ClassLoader to the Impl class is guaranteed to be the right one:

This private method is added to the Impl class:

    @SuppressWarnings({"unchecked"})
    private <T> Class<? extends T> getAsClass(
      String className, Class<? extends T> defaultClazz) {
      String sValue = className;
      if (sValue == null) {
         return defaultClazz;
      }
      try {
         return (Class<? extends T>) this.getClass().getClassLoader().loadClass(sValue);
      } catch (ClassNotFoundException e) {
         throw new NoClassSettingsException("Failed to load class setting [" 
            + className + "] with value [" + sValue + "]", e);
      }
    }

The line in question is changed to:

    Class<? extends TypeSelector> typeSelectorClass = this.getAsClass(settings.get("couchbase.typeSelector"), DefaultTypeSelector.class);

I'll commit these changes to the PR if you think they are a suitable alternative to calling the now missing Settings.getAsClass method.

Comment (Branor) [inline src/main/java/org/elasticsearch/transport/couchbase/capi/CouchbaseCAPITransportImpl.java]: Yup, that seems reasonable, please commit it to the RP.

File: pom.xml (hunk 3 above)
Comment (Branor) [inline pom.xml]: We do need to support at least 1.7, because most customers haven't moved to 1.8 yet. Supporting 1.6 is optional, because ES itself doesn't have official support for it. (https://www.elastic.co/support/matrix)

Comment (bignolip) [inline pom.xml]: Oh okay that makes sense!  I will make the adjustments to restore compatibility with 1.7.  The change is very small if I remember correctly   Thanks!

Comment (bignolip) [inline pom.xml]: I just made the change to support Java 1.7 both in this PR and in PR #105
